Dockerfile

```diff
@@ -1,3 +1,13 @@
+FROM gradle:7-jdk11-alpine AS build
+
+COPY --chown=gradle:gradle . /build
+WORKDIR /build
+
+RUN gradle shadowJar --no-daemon
+
+
 FROM openjdk:11
-COPY main/build/libs/*.jar /
+COPY --from=build /build/main/build/libs/*.jar /
 WORKDIR /
+
+ENTRYPOINT [ "java", "-jar", "gtfs-validator-0.1.0-SNAPSHOT-cli.jar" ]
```

Comment:
It looks like we don't compute the project version using [this line](https://github.com/MobilityData/gtfs-validator/blob/ef868fa2cce573f1afc22ac73b757025e17f8662/build.gradle#L33) when building via Docker.

This makes the `ENTRYPOINT` stable and works well for local testing, but creates a problem for the ones using the actual docker image. This implementation doesn't allow users to know the real validator version, which is usually displayed in the reports, which I think is very valuable. Any thoughts on this @isabelle-dr?

 If the shadow jar was created using the correct project version, we could use the file name in the `ENTRYPOINT`. @bdferris Any thoughts on this? Do you have an idea why the axion-release plugin would not work here?

Reply:
I believe if you update the actions/checkout@v2 action in the `docker.yml` file to include tag info then the version number will be computed correctly.  See https://github.com/MobilityData/gtfs-validator/blob/master/.github/workflows/package_installers.yml#L33 for an example.

Reply:
Thanks, that's helpful. Do you know if there is a way to do this locally as well? Currently, building with Docker does not calculate the version. We wouldn't need to hardcode the jar file name if it did.

Reply:
That's what I thought too. Currently, when I run this command
```
docker build . -t gtfs-validator:latest
```
I end up building a shadow jar named `gtfs-validator-0.1.0-SNAPSHOT-cli.jar`, which seems also to be the case on Chris' end since the entry point was hardcoded as such. Though, by running `./gradlew shadowJar --no-daemon` locally, the jar file is built with the correct tag version. 

Ideally, we would create an entry point using the tag detected locally and on GitHub. So I'm looking to find out why the axion-release plugin is not detected in the docker build locally. So far, I have had no luck with the documentation of the axion-release plugin on what might be the issue.

Reply:
@maximearmstrong my original version of this PR replicated what the GitHub workflow for building docker did, passing in `$VERSION_TAG` and using that to generate the entrypoint.

I removed that in 6aa301a147d54ed4f367624abb8f393f728b96fe because the jar build process stopped respecting the `$versionTag` input and started building `gtfs-validator-0.1.0-SNAPSHOT-cli.jar` regardless of input.

I found that b93dc1b75c504eac14277d8c8ba97c8f62bf9291, which got merged after I started my PR, removed that functionality so I updated this branch accordingly. We can just revert 6aa301a147d54ed4f367624abb8f393f728b96fe if the preference is to restore `$versionTag`

Reply:
Apologies, I'm new to Docker, took me a second to figure out what's going on.

1) Why wasn't the Docker build resolving the version number correctly in local builds?  Recall that the axion-release plugin uses Git tags in the cloned repo to resolve the current version.  Since `.git` was included in the `.dockerignore` file, none of the Git repro metadata was being copied over with the Docker build context (aka the `COPY --chown=gradle:gradle . /build` directive).  By removing `.git` from `.dockerignore`, I was able to build local Docker images that correctly resolved the release version number.  @themightychris did you add `.git` to `.dockerignore` for a specific reason?  I couldn't think of any reasons, but it doesn't mean they don't exist :)

2) The upshot of (1) is that gradle is now building a jar like `gtfs-validator-3.1.1-SNAPSHOT-cli.jar`.  How to specify the entry point now that the version number is no longer fixed?  I'd use something like the following:

```
COPY --from=build /build/main/build/libs/gtfs-validator-*-cli.jar /gtfs-validator-cli.jar
...
ENTRYPOINT [ "java", "-jar", "gtfs-validator-cli.jar" ]
```

Here, we are copying the gtfs-validator-VERSION-cli.jar (@themightychris note how I didn't copy *.jar, since that picks up `main-VERSION.jar` as well) and rename it to a file without a version number.  That way we can get a consistent reference in the ENTRYPOINT specifier.

I've verified the combo of (1) and (2) works for me on local Docker builds.  I think it should also work for the `docker.yml` release builds, assuming we make the `action/checkout` tweak I mentioned.

I'll admit that I'm not entirely sure what would happen if you attempted `docker build https://github.com/MobilityData/gtfs-validator.git#some_tag`.  Namely, I'm not sure if Docker will preserve enough git metadata to let version resolution work.  Worse-case scenario, it'll fail to resolve the version number, it will default back to 0.1.0-SNAPSHOT, but the ENTRYPOINT should still work correctly.

Reply:
Thanks a lot @bdferris-v2! I tested it locally and it works for me too. The correct version of the validator is displayed in the report. @themightychris If you can do these changes, we will be ready to merge.

Reply:
@bdferris-v2 It's a bit of a bad practice to pull `.git/` into the Docker build context—it can lead to extra or even sensitive content accidentally getting built into the container, and it's a criss-crossing of concerns to have your Dockerfile build process using a git client and knowledge of your release process to figure things out about the release itself.

A preferred approach is to let the system invoking the build determine the designated version and pass that in via a `--build-arg`. We can see that workflow happening previously in 6aa301a147d54ed4f367624abb8f393f728b96fe when `gradle shadowJar` could accept `versionTag` as an environment variable.

If I understand correctly, a plugin being used within the `gradle` workflow is handling extracting the current version from a local `.git/` repo? If that's the case, it would be ideal if that could be made to be overridden environmentally, such that a CI system can pass one in and deactivate the build process trying to inspect `.git/`—Would that be difficult? That's a less quick fix, but this sort of separation of build and release concerns is important for anyone who needs to run their own builds and keep their forks minimal. And it's fine if the `gradle` build layer has some smarts when its run directly as long as they can be overridden in CI via environment

Reply:
High-level: I hear what you are saying.  That said, I do wonder if we'll just be introducing complexity elsewhere for the sake of purity in the Docker file.  But more on that later.

For context, the versioning changes were relatively recent in b93dc1b75c504eac14277d8c8ba97c8f62bf9291 to better support X.Y.Z version numbers, as used in app packaging, version checks, etc.  This moved us away from the previous scheme of extracting version info from $GITHUB_REF in GitHub workflows.  One nice by-product of this change is that version info was now correctly computed for local builds as well.

So how would we make this work in a Docker build without access to .git?

First, it is possible to override the version from the Gradle command-line.  Per the [axion-release plugin docs](https://axion-release-plugin.readthedocs.io/en/latest/configuration/force_version/), you can force the version on the command-line in the following way:

```
./gradlew -Prelease.forceVersion=3.0.0
```

Second, it's possible to output the current version as deduced from .git using the Gradle [currentVersion command](https://axion-release-plugin.readthedocs.io/en/latest/configuration/tasks/#currentversion) as well:

```
./gradlew cV -q -Prelease.quiet
```

So put that all together:

1) In the docker workflow, we'd compute the current version as described above.
2) We'd pass that as a `--build-arg` to the Docker build context.
3) We'd pass that further into the actual gradle build to override the version.

Kind of roundabout, but I think it technically works?  My only question is if it's possible to gracefully fallback for local docker build such that the user doesn't have to remember to manually pass in the version argument.